src/lib/pubkey/hss_lms/lms.cpp

```diff
@@ -0,0 +1,351 @@
+                  const LMOTS_Public_Key& lmots_pk,
+                  const TreeAddress& tree_address,
+                  HashFunction& hash) {
+   hash.update(lmots_pk.identifier());
+   hash.update_be(tree_address.r());
+   hash.update_be(D_LEAF);
+   hash.update(lmots_pk.K());
+   hash.final(out);
+}
+
+std::function<void(StrongSpan<LMS_Tree_Node> out, const TreeAddress& address)> lms_gen_leaf_func(
+   const LMS_PrivateKey& lms_sk) {
+   // hash object must be shared, otherwise std::function would not be copyable, which is not allowed
+   std::shared_ptr<HashFunction> hash = HashFunction::create_or_throw(lms_sk.lms_params().hash_name());
+   return [lms_sk, hash](StrongSpan<LMS_Tree_Node> out, const TreeAddress& tree_address) {
+      auto lmots_sk = LMOTS_Private_Key(lms_sk.lmots_params(), lms_sk.identifier(), tree_address.q(), lms_sk.seed());
+      auto lmots_pk = LMOTS_Public_Key(lmots_sk);
+      lms_gen_leaf(out, lmots_pk, tree_address, *hash);
+   };
+}
+
+void lms_treehash(StrongSpan<LMS_Tree_Node> out_root,
+                  std::optional<StrongSpan<LMS_AuthenticationPath>> out_auth_path,
+                  std::optional<LMS_Tree_Node_Idx> leaf_idx,
+                  const LMS_PrivateKey& lms_sk) {
+   auto hash_pair_func = get_hash_pair_func_for_identifier(lms_sk.lms_params(), lms_sk.identifier());
+   auto gen_leaf = lms_gen_leaf_func(lms_sk);
+   TreeAddress lms_tree_address(lms_sk.lms_params().h());
+
+   treehash(out_root,
+            out_auth_path,
+            leaf_idx,
+            lms_sk.lms_params().m(),
+            LMS_TreeLayerIndex(lms_sk.lms_params().h()),
+            0,
+            hash_pair_func,
+            gen_leaf,
+            lms_tree_address);
+}
+
+}  // namespace
+
+LMS_Params LMS_Params::create_or_throw(LMS_Algorithm_Type type) {
+   uint8_t type_value = checked_cast_to_or_throw<uint8_t, Decoding_Error>(type, "Unsupported LMS algorithm type");
+
+   if(type >= LMS_Algorithm_Type::SHA256_M32_H5 && type <= LMS_Algorithm_Type::SHA256_M32_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHA256_M32_H5) + 1);
+      return LMS_Params(type, "SHA-256", h);
+   }
+   if(type >= LMS_Algorithm_Type::SHA256_M24_H5 && type <= LMS_Algorithm_Type::SHA256_M24_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHA256_M24_H5) + 1);
+      return LMS_Params(type, "Truncated(SHA-256,192)", h);
+   }
+   if(type >= LMS_Algorithm_Type::SHAKE_M32_H5 && type <= LMS_Algorithm_Type::SHAKE_M32_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHAKE_M32_H5) + 1);
+      return LMS_Params(type, "SHAKE-256(256)", h);
+   }
+   if(type >= LMS_Algorithm_Type::SHAKE_M24_H5 && type <= LMS_Algorithm_Type::SHAKE_M24_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHAKE_M24_H5) + 1);
+      return LMS_Params(type, "SHAKE-256(192)", h);
+   }
```

Comment:
I'm both scared and amazed. But frankly, I think enum-arithmetics is married to pointer-arithmetics and they made a bunch of children called "obscure bug" and "undefined behavior". 🤡

Would a `switch()` in some helper function really be so cumbersome? 😨

Reply:
If you insist 😝

Reply:
Thank you.